contracts/zero-ex/contracts/src/features/MetaTransactions.sol

```diff
@@ -0,0 +1,412 @@
+                    state.sender,
+                    state.mtx.sender
+                ).rrevert();
+        }
+        // Must not be expired.
+        if (state.mtx.expirationTime <= block.timestamp) {
+            LibMetaTransactionsRichErrors
+                .MetaTransactionExpiredError(
+                    state.hash,
+                    block.timestamp,
+                    state.mtx.expirationTime
+                ).rrevert();
+        }
+        // Must have a valid gas price.
+        if (state.mtx.minGasPrice > tx.gasprice || state.mtx.maxGasPrice < tx.gasprice) {
+            LibMetaTransactionsRichErrors
+                .MetaTransactionGasPriceError(
+                    state.hash,
+                    tx.gasprice,
+                    state.mtx.minGasPrice,
+                    state.mtx.maxGasPrice
+                ).rrevert();
+        }
+        // Must have enough ETH.
+        state.selfBalance  = address(this).balance;
+        if (state.mtx.value > state.selfBalance) {
+            LibMetaTransactionsRichErrors
+                .MetaTransactionInsufficientEthError(
+                    state.hash,
+                    state.selfBalance,
+                    state.mtx.value
+                ).rrevert();
+        }
+        // Must be signed by signer.
+        try
+            ISignatureValidator(address(this))
+                .validateHashSignature(state.hash, state.mtx.signer, state.signature)
+        {}
+        catch (bytes memory err) {
+            LibMetaTransactionsRichErrors
+                .MetaTransactionInvalidSignatureError(
+                    state.hash,
+                    state.signature,
+                    err
+                ).rrevert();
+        }
+        // Transaction must not have been already executed.
+        state.executedBlockNumber = LibMetaTransactionsStorage
+            .getStorage().mtxHashToExecutedBlockNumber[state.hash];
+        if (state.executedBlockNumber != 0) {
+            LibMetaTransactionsRichErrors
+                .MetaTransactionAlreadyExecutedError(
+                    state.hash,
+                    state.executedBlockNumber
+                ).rrevert();
+        }
+    }
+
+    /// @dev Execute a `ITransformERC20.transformERC20()` meta-transaction call
+    ///      by decoding the call args and translating the call to the internal
+    ///      `ITransformERC20._transformERC20()` variant, where we can override
+    ///      the taker address.
+    function _executeTransformERC20Call(ExecuteState memory state)
+        private
+        returns (bytes memory returnData)
+    {
+        // HACK(dorothy-zbornak): `abi.decode()` with the individual args
+        // will cause a stack overflow. But we can prefix the call data with an
+        // offset to transform it into the encoding for the equivalent single struct arg.
+        // Decoding a single struct consumes far less stack space.
+        TransformERC20Args memory args;
+        {
```

Comment:
I think this looks okay but I'll have to spend more time validating it. Could we add more documentation around the expected memory layouts? See [here](https://github.com/0xProject/0x-monorepo/blob/development/contracts/asset-proxy/contracts/src/ERC20Proxy.sol#L67) as an example.